File: Problems/542_01_Matrix.py
from typing import List
from collections import deque

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        ROW = len(mat)
        COL = len(mat[0])
        dist = [[0 for c in range(COL)]for r in range(ROW)]
        curr = deque([])
        for r in range(ROW):
            for c in range(COL):
                if mat[r][c]:
                    dist[r][c] = 10**4+1
                else:
                    curr.append((r, c))
        directions = ((0, 1), (1, 0), (-1, 0), (0, -1))
        while curr:
            curr_r, curr_c = curr.popleft()
            for dir in directions:
                nxt_r, nxt_c = curr_r + dir[0], curr_c + dir[1]
                if ROW>nxt_r>=0 and COL>nxt_c>=0 and dist[nxt_r][nxt_c]>dist[curr_r][curr_c]+1:
                        dist[nxt_r][nxt_c] = dist[curr_r][curr_c]+1
                        curr.append((nxt_r, nxt_c))
        return dist

File: Problems/542_01_Matrix.cpp
#include<vector>
#include<queue>

using namespace std;

class Solution {
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int ROW = mat.size();
        int COL = mat[0].size();
        vector<vector<int>> dist(ROW, vector<int>(COL, __INT_MAX__));
        queue<pair<int, int>> qu;
        for(int r=0; r<ROW; r++){
            for(int c=0; c<COL; c++){
                if(mat[r][c]==0){
                    dist[r][c] = 0;
                    qu.push({r, c});
                }
            }
        }
        vector<pair<int, int>> directions = {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
        while(!qu.empty()){
            int curr_r = qu.front().first;
            int curr_c = qu.front().second;
            qu.pop();
            for(auto dir: directions){
                int nxt_r = curr_r + dir.first;
                int nxt_c = curr_c + dir.second;
                if(ROW>nxt_r && nxt_r>=0 && COL>nxt_c && nxt_c>=0 && dist[nxt_r][nxt_c]>dist[curr_r][curr_c]+1){
                    dist[nxt_r][nxt_c] = dist[curr_r][curr_c]+1;
                    qu.push({nxt_r, nxt_c});
                }
            }
        }
        return dist;
    }
};